File: tvn-cosine.ai/tvn-cosine.ai/environment/wumpusworld/ManhattanHeuristicFunction.cs
using tvn.cosine.ai.agent.api;
using tvn.cosine.ai.common.collections;
using tvn.cosine.ai.search.framework;
using tvn.cosine.ai.util;

namespace tvn.cosine.ai.environment.wumpusworld
{
    /**
     * Heuristic for calculating the Manhattan distance between two rooms within a Wumpus World cave. 
     */
    public class ManhattanHeuristicFunction : ToDoubleFunction<Node<AgentPosition, IAction>>
    {
        IQueue<Room> goals = Factory.CreateQueue<Room>();

        public ManhattanHeuristicFunction(ISet<Room> goals)
        {
            this.goals.AddAll(goals);
        }

        public double applyAsDouble(Node<AgentPosition, IAction> node)
        {
            AgentPosition pos = node.getState();
            int nearestGoalDist = int.MaxValue;
            foreach (Room g in goals)
            {
                int tmp = evaluateManhattanDistanceOf(pos.getX(), pos.getY(), g.getX(), g.getY());

                if (tmp < nearestGoalDist)
                {
                    nearestGoalDist = tmp;
                }
            }

            return (double)nearestGoalDist;
        }

        private int evaluateManhattanDistanceOf(int x1, int y1, int x2, int y2)
        {
            return System.Math.Abs(x1 - x2) + System.Math.Abs(y1 - y2);
        }
    }
}

File: tvn-cosine.ai/tvn-cosine.ai.v2/environment/wumpusworld/ManhattanHeuristicFunction.java
namespace aima.core.environment.wumpusworld;

using java.util.ArrayList;
using java.util.List;
using java.util.Set;
using java.util.function.ToDoubleFunction;

using aima.core.environment.wumpusworld.action.WWAction;
using aima.core.search.api.Node;

/**
 * Heuristic for calculating the Manhattan distance between two rooms within a
 * Wumpus World cave.
 * 
 * @author Federico Baron
 * @author Alessandro Daniele
 * @author Ciaran O'Reilly
 * @author Anurag Rai
 */
public class ManhattanHeuristicFunction implements ToDoubleFunction<Node<WWAction, AgentPosition>> {

	List<Room> goals = new ArrayList<>();

	public ManhattanHeuristicFunction(ISet<Room> goals) {
		this.goals.addAll(goals);
	}

	 
	public double applyAsDouble(Node<WWAction, AgentPosition> value) {
		return value.pathCost() + h(value.state());
	}

	private double h(AgentPosition state) {
		AgentPosition pos = state;
		int nearestGoalDist = Integer.MAX_VALUE;
		for (Room g : goals) {
			int tmp = evaluateManhattanDistanceOf(pos.getX(), pos.getY(), g.getX(), g.getY());

			if (tmp < nearestGoalDist) {
				nearestGoalDist = tmp;
			}
		}

		return nearestGoalDist;
	}

	//
	// PRIVATE
	//
	private int evaluateManhattanDistanceOf(int x1, int y1, int x2, int y2) {
		return Math.abs(x1 - x2) + Math.abs(y1 - y2);
	}
}
